File: Solutions/593_Valid_Square/593_Valid_Square.py
class Solution:
    def validSquare(self, p1: List[int], p2: List[int], p3: List[int], p4: List[int]) -> bool:
        def dist(p,q):
            return (p[0]-q[0])**2 + (p[1]-q[1])**2
        
        def check(p1,p2,p3,p4):
            if p1[0]+p2[0] != p3[0]+p4[0] or p1[1]+p2[1] != p3[1]+p4[1]: return False
            if dist(p1,p2) != dist(p3,p4) or dist(p1,p4) != dist(p2,p4): return False
            if p1 == p2: return False
            return True
        
        return check(p1,p2,p3,p4) or check(p1,p3,p2,p4) or check(p1,p4,p2,p3)

#Third fastest
import itertools

class Solution:
    
    
    def get_multiplicities(self, points):
        
        distances = {}
        
        for combination in itertools.combinations(points, 2):
            distance = math.sqrt((combination[0][0] - combination[1][0]) ** 2 + (combination[0][1] - combination[1][1]) ** 2)
            if distance not in distances.keys():
                distances[distance] = 0
            if len(distances) > 2:
                return False
            distances[distance] += 1
        
        return len(distances) == 2 and 2 in distances.values() and 4 in distances.values()
    
    def validSquare(self, p1: List[int], p2: List[int], p3: List[int], p4: List[int]) -> bool:
        return self.get_multiplicities([p1, p2, p3, p4])

#Second fastest
class Solution:
    def validSquare(self, p1: List[int], p2: List[int], p3: List[int], p4: List[int]) -> bool:
        def dist(P, Q):
            return (P[0] - Q[0])**2 + (P[1] - Q[1])**2

        D = [dist(p1, p2), dist(p1, p3), dist(p1, p4),
             dist(p2, p3), dist(p2, p4), dist(p3, p4)]
        D.sort()
        return 0 < D[0] == D[1] == D[2] == D[3] and 2*D[0] == D[4] == D[5]

#Fastest
class Solution:
    def distance(self,p1,p2):
        x1,y1 = p1
        x2,y2 = p2 
        return math.sqrt(((x2 - x1) ** 2) + ((y2 - y1) ** 2))
        
        
    def validSquare(self, p1: List[int], p2: List[int], p3: List[int], p4: List[int]) -> bool:
          
        # if p1==p2 or p1==p3 or p1==p4 or p2==p3 or p2==p4 or p3==p4:
        #     return False

        points = [[p1,p2],[p1,p3],[p1,p4],[p2,p3],[p2,p4],[p3,p4]]
        res=[]

        for i in points:
            res.append(self.distance(i[0],i[1]))
            
        h={}

        for i in res:
            if i not in h:
                h[i] = 1
            else:
                h[i]+=1
        for i in h.values():
            if (i==2 or i==4) and len(h)==2:
                continue
            else:
                return False
        return True

        # if len(set(res))==2:
        #     return True
        # else:
        #     return False

File: Solutions/593_Valid_Square/593_Valid_Square.cpp
//Second fastest
class Solution {
public:
    bool validSquare(vector<int>& p1, vector<int>& p2, vector<int>& p3, vector<int>& p4) {
        int dis_1_to_2 = distance2power(p1, p2);
        int dis_1_to_3 = distance2power(p1, p3);
        int dis_1_to_4 = distance2power(p1, p4);

        if (dis_1_to_2 == 0 || dis_1_to_3 == 0 || dis_1_to_4 == 0) {
            return false;
        } else if (dis_1_to_2 == dis_1_to_3 && dis_1_to_4 == (dis_1_to_2 + dis_1_to_3)) {
            // judge slope
            return is90degree(p2, p1, p3) && is90degree(p2, p4, p3);
        } else if (dis_1_to_2 == dis_1_to_4 && dis_1_to_3 == (dis_1_to_2 + dis_1_to_4)) {
            return is90degree(p2, p1, p4) && is90degree(p2, p3, p4);
        } else if (dis_1_to_3 == dis_1_to_4 && dis_1_to_2 == (dis_1_to_3 + dis_1_to_4)) {
            return is90degree(p3, p1, p4) && is90degree(p3, p2, p4);
        }
        return false;
    }

    int distance2power(vector<int> &a, vector<int> &b) {
        return abs((a[0] - b[0]) * (a[0] - b[0]) + (a[1] - b[1]) * (a[1] - b[1]));
    }

    double is90degree(vector<int> &a, vector<int> &b, vector<int> &c) {
        return (b[1] - a[1]) * (b[1] - c[1]) + (b[0] - a[0]) * (b[0] - c[0]) == 0;
    }
};

//Another
#include<math.h>
class Solution {
public:
    double calc(vector<int>& p1, vector<int>& p2){
        return sqrt((p1[0]-p2[0])*(p1[0]-p2[0])+(p1[1]-p2[1])*(p1[1]-p2[1]));
    }
    bool validSquare(vector<int>& p1, vector<int>& p2, vector<int>& p3, vector<int>& p4) {
        set<double> s;
        if(p1==p2 || p2==p3 || p3==p4 || p4==p1)return false;
        double d1=calc(p1,p2),d2=calc(p2,p3),d3=calc(p3,p4),d4=calc(p4,p1),d5=calc(p1,p3),d6=calc(p2,p4);
        s.insert(d1);s.insert(d2);s.insert(d3);s.insert(d4);s.insert(d5);s.insert(d6);
        return s.size()==2;
    }
};

//Fastest
class Solution {
public:
    double d(vector<int>& p1, vector<int>& p2) {
        return pow(pow(p1[0] - p2[0], 2) + pow(p1[1] - p2[1], 2), 0.5);
    }
    bool validSquare(vector<int>& p1, vector<int>& p2, vector<int>& p3, vector<int>& p4) {
        // support variables
        double d0, d1, d2, d3;
        vector<vector<int>> pts = {p1, p2, p3, p4};
        sort(begin(pts), end(pts));
        // computing the distances between ordered points
        d0 = d(pts[0], pts[1]), d1 = d(pts[0], pts[2]), d2 = d(pts[0], pts[3]), d3 = d(pts[1], pts[2]);
        // checking if sides are non-0 and if sides and diagonals are equal
        return d0 && d2 && d0 == d1 && d2 == d3;
    }
};

//Another
class Solution {
public:
    int getSquaredDistance(vector<int> &a, vector<int> &b) {
        int dx = a[0]-b[0];
        int dy = a[1]-b[1];
        return dx*dx + dy*dy;
    }
    
    bool validSquare(vector<int>& p1, vector<int>& p2, vector<int>& p3, vector<int>& p4) {
        // connected, perpecular
        
        vector<int> dist_sq(6);
        dist_sq[0] = getSquaredDistance(p1, p2);
        dist_sq[1] = getSquaredDistance(p1, p3);
        dist_sq[2] = getSquaredDistance(p1, p4);
        dist_sq[3] = getSquaredDistance(p2, p3);
        dist_sq[4] = getSquaredDistance(p2, p4);
        dist_sq[5] = getSquaredDistance(p3, p4);
        sort(dist_sq.begin(), dist_sq.end());
        
        //For the Six Distances in Sorted Array
        //The 1st and the 4th distance should be equal (This implies first 4 are equal)
        //The 5th and the 6th distance should be equal
        //Side should not be zero/Distance of side and diagonal shouldn't be equal
        return (dist_sq[0]>0  // >0 for all
                && dist_sq[0]==dist_sq[3] // first 4 equal
                && dist_sq[4]==dist_sq[5] // longest 2 (diagnal) equal
                && dist_sq[3]<dist_sq[4]); // side < diagnal
    }
};